applications/MeshMovingApplication/custom_utilities/fixed_mesh_ale_utilities.cpp

```diff
@@ -0,0 +1,532 @@
+
+        // Set a linear strategy to solve the mesh moving problem
+        const unsigned int echo_level = 2;
+        const bool compute_reactions = false;
+        const bool calculate_norm_dx_flag = false;
+        const bool reform_dof_set_at_each_step = false;
+        mpMeshMovingStrategy = Kratos::make_shared<StrategyType>(
+            mrVirtualModelPart,
+            p_scheme,
+            mpLinearSolver,
+            p_builder_and_solver,
+            compute_reactions,
+            reform_dof_set_at_each_step,
+            calculate_norm_dx_flag);
+
+        mpMeshMovingStrategy->Check();
+        mpMeshMovingStrategy->Initialize();
+        mpMeshMovingStrategy->SetEchoLevel(echo_level);
+    }
+
+    const Vector FixedMeshALEUtilities::SetDistancesVector(ModelPart::ElementIterator ItElem)
+    {
+        auto &r_geom = ItElem->GetGeometry();
+        Vector nodal_distances(r_geom.PointsNumber());
+
+        if (mLevelSetType == "continuous"){
+            // Continuous nodal distance function case
+            for (unsigned int i_node = 0; i_node < r_geom.PointsNumber(); ++i_node) {
+                nodal_distances[i_node] = r_geom[i_node].FastGetSolutionStepValue(DISTANCE);
+            }
+        } else if (mLevelSetType == "discontinuous") {
+            // Discontinuous elemental distance function case
+            nodal_distances = ItElem->GetValue(ELEMENTAL_DISTANCES);
+        } else {
+            KRATOS_ERROR << "Level set type must be either 'continuous' or 'discontinuous'. Got " << mLevelSetType;
+        }
+
+        return nodal_distances;
+    }
+
+    inline bool FixedMeshALEUtilities::IsSplit(const Vector &rDistances)
+    {
+        unsigned int n_pos = 0, n_neg = 0;
+        for (double dist : rDistances) {
+            if(dist >= 0) {
+                ++n_pos;
+            } else {
+                ++n_neg;
+            }
+        }
+        if (n_pos > 0 && n_neg > 0) {
+            return true;
+        }
+        return false;
+    }
+
+    void FixedMeshALEUtilities::InitializeVirtualMeshValues()
+    {
+        // Copy the PRESSURE and VELOCITY values from the origin mesh to the virtual one
+        VariableUtils::Pointer p_var_utils = Kratos::make_shared<VariableUtils>();
+        for (unsigned int i_step = 0; i_step <  mpOriginModelPart->GetBufferSize(); ++i_step) {
+            p_var_utils->CopyModelPartNodalVar<Variable<double>>(PRESSURE, *mpOriginModelPart, mrVirtualModelPart, i_step);
+            p_var_utils->CopyModelPartNodalVar<Variable<array_1d<double,3>>>(VELOCITY, *mpOriginModelPart, mrVirtualModelPart, i_step);
+        }
+
+        // Initialize the DISPLACEMENT, MESH_DISPLACMENT and MESH_VELOCITY values
+        p_var_utils->SetHistoricalVariableToZero(MESH_VELOCITY, mrVirtualModelPart.Nodes());
+        p_var_utils->SetHistoricalVariableToZero(MESH_DISPLACEMENT, mrVirtualModelPart.Nodes());
+    }
+
+    void FixedMeshALEUtilities::InitializeMeshDisplacementFixity()
```

Comment:
I think there is a function in the VariableUtils that does exatly this

Reply:
Thanks! I'll check it!

Reply:
In this case is faster to do it in this way since we free the three DOFs at the same time. The one in the VariableUtils would imply to free the components one by one, so to loop the nodes container 3 times (X,Y and Z).